sdks/java/core/src/main/java/org/apache/beam/sdk/io/WriteFiles.java

```diff
@@ -133,37 +133,50 @@
   private final ValueProvider<Integer> numShardsProvider;
   private final boolean windowedWrites;
   private int maxNumWritersPerBundle;
+  // If set,this is used instead of DynamicDestinations.getSideInputs. This is used by the Beam
+  // translation code.
+  private final List<PCollectionView<?>> overriddenSideInputs;
+
+  private List<PCollectionView<?>> getSideInputs() {
+    return (overriddenSideInputs != null)
+        ? overriddenSideInputs
+        : getSink().getDynamicDestinations().getSideInputs();
```

Comment:
Big point: The core SDK should be translation unaware. So comments about how it is to support translation are not appropriate.

In fact, this is to support pipeline surgery, not translation. Surgery is where PCollections get swapped out arbitrarily, so you can't just ask the UDF what the side inputs are, but need to match up the tags it claims to have with the actual inputs to the transform, which can change while the UDF stays the same.

It is also explicitly supported/encouraged for a user to apply a `PTransform` more than once (or apply `WriteFiles` to the same sink more than once) so rewiring the side inputs seems like a natural top level thing for a user.

Technically, I want to provide context for how I am thinking about this.

You have not chosen to be like `ParDo`, which is:

 - The `UDF` has `PCollectionView` embedded in its closure and used just as a tag. But it is not actually aware of which ones it has.
 - Separately, the same `PCollection` is wired into the graph via `.withSideInputs`. In this way, both the graph and the UDF know about them.

Instead, your model is:

 - The UDF is aware of the `PCollectionViews` that are embedded, so it can tell you what they are.
 - The transform just asks the UDF what the side inputs are. This can only happen during construction, since a UDF in another language cannot be invoked. 

There are a couple of caveats to this latter approach:

 - Even if the UDF knows the whole `PCollectionView` it must still use them only as tags, since the `PCollection` itself is free to be replaced during surgery.
 - Because the transform's inputs can be rewired at will, it can ask `sink.getSideInputs()` but it is semantically required to take those tags and match them against its actual side inputs. It is a special case if the transform knows that its actual side inputs are the ones that the UDF claims.

I think what you have is approximately implementing this, but this conditional belongs in the constructor.

Aesthetically, I would expect more pitfalls to this tighter coupling of the Java bits and pieces and the graph structure, but I don't have anything more off the top of my head.